core/block_validator.go

```diff
@@ -131,7 +132,39 @@ func (v *BlockValidator) ValidateState(block *types.Block, statedb *state.StateD
 	if root := statedb.IntermediateRoot(v.config.IsEIP158(header.Number)); header.Root != root {
 		return fmt.Errorf("invalid merkle root (remote: %x local: %x) dberr: %w", header.Root, root, statedb.Error())
 	}
-	// Verify that the advertised root is correct before
+	// In verkle mode, verify the proof.
+	if v.config.IsVerkle(header.Number, header.Time) {
+		// Check that all keys in the witness were used
+		keys := statedb.Witness().Keys()
+		var key [32]byte
+		for _, stemdiff := range block.ExecutionWitness().StateDiff {
+			copy(key[:31], stemdiff.Stem[:])
+			for _, suffixdiff := range stemdiff.SuffixDiffs {
+				key[31] = suffixdiff.Suffix
+
+				var found bool
+				for _, k := range keys {
+					if bytes.Equal(k, key[:]) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return fmt.Errorf("superfluous key %x could not be found in witness", key)
+				}
+			}
+		}
```

Comment:
I'm only checking that the leaves from the block witness are present in the destination witness, because we are targeting state bloat. It could be that the witness does _not_ contain all necessary leaves for stateless execution. This should be checked as well, either by execution (slow) or by just checking that there is the exact same amount of leaves.